modules/distributor/forwarder/manager.go

```diff
@@ -66,21 +66,49 @@ func NewManager(cfgs ConfigList, logger log.Logger, overrides Overrides) (*Manag
 }
 
 func (m *Manager) ForTenant(tenantID string) List {
-	m.tenantToQueueListMu.RLock()
-	defer m.tenantToQueueListMu.RUnlock()
-
-	ql, found := m.tenantToQueueList[tenantID]
-	if !found {
+	ql, ok := m.getOrCreateQueueList(tenantID)
+	if !ok {
 		return nil
 	}
 
 	return ql.list
 }
 
-func (m *Manager) start(_ context.Context) error {
-	tenantToForwarderNames := m.readOverrides()
-	m.updateTenantToForwarderList(tenantToForwarderNames)
+func (m *Manager) getOrCreateQueueList(tenantID string) (*queueList, bool) {
+	ql, found := m.getQueueList(tenantID)
+	if found {
+		return ql, true
+	}
+
+	m.tenantToQueueListMu.Lock()
+	defer m.tenantToQueueListMu.Unlock()
+
+	forwarderNames := m.overrides.Forwarders(tenantID)
```

Comment:
it seems like for tenants w/o a configured forwarder this lock will hit repeatedly b/c it will not be found in the list. should we store a `nil` value in `m.tenantToQueueList` to prevent this?

Reply:
It seems like this is the core of the change? By repeatedly asking for the forwarder we can catch the wildcard scenarios? I'm fine with that idea, but I'd like to find a way to "cache" the nil to prevent constant lookups here.

Other than that looks good.

Reply:
I see now. I'll admit I don't like the logic to get the queue is split between this function and  `distributor.go:382`. It's difficult to see all the logic related to how we request a forwarder for a tenant. Likely it was put there because it mirrors the metric generator lines above it?

Personally, I think it would be cleaner to move this logic:
```
	if len(d.cfg.Forwarders) > 0 && len(d.overrides.Forwarders(userID)) > 0 {
```

into the manager b/c there would be one place to look to see exactly how a forwarder is requested for a tenant. This is not a blocker though. WDYT?

Reply:
Moved in [ce63eeb](https://github.com/grafana/tempo/pull/1871/commits/ce63eebee0fd54436bcaf1f16fa5c2d1c862f067).